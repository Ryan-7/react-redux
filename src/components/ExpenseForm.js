import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';


const now = moment();
console.log(now.format('MMM Do, YYYY'));

export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        console.log("the props the form received in this instance:")
        console.log(props)
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),         // the date/time when the page is loaded
            calendarFocused: false,
            error: ''
        }
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({
            description: description
        }));
    }

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({
            note: note
        }));
    }

    onAmountChange = (e) => {
        const amount = e.target.value;
        
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({
                amount: amount
            }))
        }
    }

    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({
                createdAt: createdAt
            }))
        }
    }

    onFocusChange = ({ focused }) => {
        this.setState(() => ({
            calendarFocused: focused
        }))
    }

    onSubmit = (e) => {
        e.preventDefault();

        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({
                error: 'Please provide description and amount'
            }));
        } else {
            this.setState(() => ({
                error: ''
            }));

            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100, // format the number into just cents (no decimals)
                createdAt: this.state.createdAt.valueOf(), // use moment to set date to a number (milliseconds since epoch)
                note: this.state.note,
            });
        }
    }

    // onRemoveExpense = () => {
    //   //  this.props.removeExpense();
    //   console.log('works')
    // }

    render() {
        return (
            <div>
                <form onSubmit={this.onSubmit}>
                    <input 
                        type="text" 
                        placeholder="Description" 
                        autoFocus 
                        value={this.state.description} 
                        onChange={this.onDescriptionChange} 
                    />
                    <input 
                        type="text" 
                        placeholder="Amount" 
                        value={this.state.amount}
                        onChange = {this.onAmountChange}
                    />
                    <SingleDatePicker 
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false} // pick days in the past / future 
                    
                    />

                    <textarea 
                        placeholder="Add a note for your expense (optional)"
                        value={this.state.note}
                        onChange={this.onNoteChange} 
                    >
                    </textarea>
                    <button>Add Expense</button>
                </form>
                {this.state.error && <p>{this.state.error}</p>}
                <button hidden={!this.props.expense} onClick={this.props.removeExpense}>Remove</button>
            </div>
        )
    }
}


