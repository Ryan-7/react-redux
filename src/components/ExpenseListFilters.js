import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import moment from 'moment';


import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';

class ExpenseListFilters extends React.Component {
    constructor(props) {
        super(props)

        console.log(this.props.filters.startDate)
        console.log(this.props.filters.endDate)
    }

    state = {
        calendarFocused: null
    }

    onDatesChange = ({ startDate, endDate }) => { // these arguments are provided by the library when we select different dates 
         this.props.dispatch(setStartDate(startDate));
         this.props.dispatch(setEndDate(endDate));
    }

    onFocusChange = (calendarFocused) => {
        this.setState(() => {
            return  { 
                calendarFocused: calendarFocused 
            }    
        })
    }

    render() {
        return (
            <div>
                <input type='text' value={this.props.filters.text} onChange={(e) =>{
                    console.log(e.target.value);
                    this.props.dispatch(setTextFilter(e.target.value))
                }} />
                <select
                    value={this.props.filters.sortBy} 
                    onChange={(e) => {
                        if (e.target.value === 'date') {
                            this.props.dispatch(sortByDate());
                        } else if (e.target.value === 'amount') {
                            this.props.dispatch(sortByAmount());
                        }
                }}>
                    <option></option>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker 
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                    showClearDates={true}
                />
            </div>
        )
    }
}


const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
}

export default connect(mapStateToProps)(ExpenseListFilters);


