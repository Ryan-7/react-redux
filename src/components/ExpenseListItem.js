import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

// props is one giant object, react will combine it all for you:
// the props from the parent, and the props from connect (dispatch, etc), props from the store, etc.
// props are props

const ExpenseListItem = ({ description, amount, createdAt, id }) => (
    <div>
        <Link to={`/edit/${id}`}><h3>{description}</h3></Link>
        <p>
            {numeral(amount / 100).format('$0,0.00')}
            - 
            {console.log(createdAt)}
            {moment(createdAt).format('MMMM Do, YYYY')}
        </p>

    </div>
)

// No need for the mapStateToProps function if we're not using the state.
// Want to connect so we can access dispatch

export default ExpenseListItem;


