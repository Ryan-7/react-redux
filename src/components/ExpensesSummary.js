import React from 'react';
import { connect } from 'react-redux'; // allows us to connect to the store
import expensesTotal from '../selectors/expensesTotal';
import getVisibleExpeneses from '../selectors/expenses'
import numeral from 'numeral';

export const ExpensesSummary = ({ numOfExpenses, totalAmounts }) => (
    <div>
        <p>Expenses Summary</p>
        <p>Viewing {numOfExpenses} {numOfExpenses === 1 ? 'expense' : 'expenses'} totaling {numeral(totalAmounts / 100).format('$0,0.00')}</p>
    </div>
)


const mapStateToProps = (state) => {
    return {
        totalAmounts: expensesTotal(getVisibleExpeneses(state.expenses, state.filters)),
        numOfExpenses: getVisibleExpeneses(state.expenses, state.filters).length
    }
}

export default connect(mapStateToProps)(ExpensesSummary);