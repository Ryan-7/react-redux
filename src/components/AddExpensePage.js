import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { addExpense } from '../actions/expenses';


export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.addExpense(expense);
        this.props.history.push('/');
    };

    render() {
        return (
            <div>
                <h1>Add Expense</h1>
                <ExpenseForm
                    onSubmit={this.onSubmit} 
                />
            </div>
        )
    }

}


// A way to extract the dispatch functions away from the component itself 

const mapDispatchToProps = (dispatch) => {
    return {
        addExpense: (expense) => dispatch(addExpense(expense))
    };
}

export default connect(undefined, mapDispatchToProps)(AddExpensePage);







// const AddExpensePage = (props) => (
//     <div>
//         <h1>Add Expense</h1>
//         <ExpenseForm onSubmit={(expense) => {
//             console.log(expense);
//             props.dispatch(addExpense(expense));
//             props.history.push('/');
//         }}/>
//     </div>
// )

// export default connect()(AddExpensePage);
