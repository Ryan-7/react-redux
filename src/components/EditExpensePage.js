import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense } from '../actions/expenses';
import { removeExpense } from '../actions/expenses';

const EditExpensePage = (props) => {
    return (
        <div>
            <ExpenseForm 
                expense={props.expense} // the expense we obtained from the store according to the id 
                onSubmit={(expense) => {
                    props.dispatch(editExpense(props.expense.id, expense));
                    props.history.push('/');
                }}
                removeExpense={(t) => {
                    props.dispatch(removeExpense({id: props.expense.id}));
                    props.history.push('/');
                }}
            />
        </div>
    )
}


// Can also access the props in the regular component in the HOC component as well, so we can get the param id for example
// Remember, what is returned here is just amended to our component above and can be accessed through props

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => {
            return expense.id === props.match.params.id
        })
    }
}

export default connect(mapStateToProps)(EditExpensePage);