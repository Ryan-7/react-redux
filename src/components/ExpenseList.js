import React from 'react';
import { connect } from 'react-redux'; // allows us to connect to the store
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';


const ExpenseList = (props) => (
    <div>
        <h1>Expense List</h1>
        {props.expenses.map((item) => {
           // return <ExpenseListItem description={item.description} amount={item.amount} createdAt={item.createdAt}/>
           return <ExpenseListItem key={item.id} {...item} />
        })}
    </div>
);

// As the store changes, the component will re-render, it's reactive
// state gives us access to anything on our store 
// this object is appended to our props and can be used 

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters), // this now exists on our component as props.expenses, we are enhancing our component
        stuff: state.nickname
    }
}

export default connect(mapStateToProps)(ExpenseList);



// const ConnectedExpenseList = connect((state) => {
//     return {
//         expenses: state.expenses
//     }
// })(ExpenseList);

// export default ConnectedExpenseList;