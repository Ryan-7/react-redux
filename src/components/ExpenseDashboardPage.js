import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

const ExpenseDashboardPage = () => {
    return (
        <div>
            <ExpenseListFilters />
            <ExpenseList />
        </div>
    )
}




// Use this syntax for Stateless Functional Components
export default ExpenseDashboardPage;


// If it was a ES6 class component, use this syntax:
// export default class ExpenseDashboardPage extends React.Component {}