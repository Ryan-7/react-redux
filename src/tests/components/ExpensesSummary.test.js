import React from 'react';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import { shallow } from 'enzyme';

test('should correctly render ExpensesSummary with 1 expense', () => {
    const wrapper = shallow(<ExpensesSummary numOfExpenses={1} totalAmounts={100} />);
    expect(wrapper).toMatchSnapshot();

})

test('should correctly render ExpensesSummary with multiple expenses', () => {
    const wrapper = shallow(<ExpensesSummary numOfExpenses={23} totalAmounts={100} />);
    expect(wrapper).toMatchSnapshot();
})