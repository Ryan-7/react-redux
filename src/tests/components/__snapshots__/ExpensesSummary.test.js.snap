// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render ExpensesSummary with 1 expense 1`] = `
<div>
  <p>
    Expenses Summary
  </p>
  <p>
    Viewing 
    1
     
    expense
     totaling 
    $1.00
  </p>
</div>
`;

exports[`should correctly render ExpensesSummary with multiple expenses 1`] = `
<div>
  <p>
    Expenses Summary
  </p>
  <p>
    Viewing 
    23
     
    expenses
     totaling 
    $1.00
  </p>
</div>
`;
